### ISSUE TITLE
Implement a simpler way to express lists with a separator

### ISSUE THREAD

**Issue by lontivero** [pegjs/pegjs#107]
When we have a right recursion we have to do something like this:

``` javascript
Statements
  = head:Statement tail:(__ Statement)* {
      var result = [head];
      for (var i = 0; i < tail.length; i++) {
        result.push(tail[i][1]);
      }
      return result;
    }
```

I could find useful to be able to do the same as follow:

``` javascript
Statements
  = st:Statement (__ st:Statement)* { return st; /*where st is an array*/ }
```

**Comment by alexchandel** [comment]
Perl's [Regexp::Grammars](search.cpan.org/perldoc/Regexp::Grammars) does this with the modulus operator as well:

```
# a list of one or more "item", separated by "separator", returning an array:
<rule: list>
        <[item]>+ % <separator>
```

It's sort of an arguable use of the modulus. The floating point modulus defines a quotient group, a half-open interval of reals (IEEE 754 floats really). It's a sloppy analogy since the items returned by the Regexp::Grammar modulus are only identical up to a pattern (and more importantly since strings aren't a group) but it's close enough.

**Comment by polkovnikov-ph** [comment]
Instead of making it a built-in operator, I've just made parsers that could be parameterized by parsers.
[Metagrammar here.](https://github.com/polkovnikov-ph/newpeg/blob/master/parse.np)

**Comment by futagoza** [comment]
The most related one I could find is #45 but the OP of that issue proposes a different syntax (similar to what you've implemented in your metagrammar @polkovnikov-ph), but I plan to go with parametric rules (like @dmajda suggested above) that use the more common syntax of `< .. >` (templates, generics, etc).

**Comment by polkovnikov-ph** [comment]
@futagoza Syntax doesn't really matter. Whatever you end up doing with this issue, I greatly appreciate that.

**Comment by StoneCypher** [comment]
The choice of symbol matters a lot, as this can interfere with other high priority features, like @Mingun 's range patch 

There isn't a particularly sensible alternative for ranges, so my inclination would be to preserve angle braces for those

Being honest I'm kind of a parser noob.  My solution to this seems simple and I'm a little worried that it might not work at scale, but why not do something like

```
fname = "bob" / "dan" / "charlie"
namelist = (WS? fname ","?)+
```

If there's an actual need here, I support it - lists are amongst the most common things a parser has to do, and it seems like there probably is an actual need here, because several strong users are in this thread and didn't say that

So my solution probably isn't good enough, but I'd like to learn why

**Comment by Mingun** [comment]
> So my solution probably isn't good enough, but I'd like to learn why

Mostly because it accepts some inputs, that in real life must be unacceptable. For example, it parses `bobdan`.

**Comment by StoneCypher** [comment]
Fair.  That was na&iuml;ve of me.

This allows `bob`, `dan`, and `bob,dan`, but not `bobdan`.  What did I miss here?

```
Document = names
WS = [ \r\n]+

fname = "bob" / "dan" / "charlie"

nameitem = (WS? fname ",")+ 
namelastitem = (WS? fname)

namelist = nameitem? namelastitem
```

**Comment by StoneCypher** [comment]
@polkovnikov-ph - there's also #36, which is similar to #45 but not identical

The author's explanation seems to suggest that 36 is closer to what you meant

**Comment by Mingun** [comment]
Now it parses only what is expected, but results not is a single array. This is main motivation to would have special construct to parse delimited data

**Comment by StoneCypher** [comment]
Okay, this is starting to look like something that genuinely should have an operator, as an ease of use thing.  Most users won't have a Russian expert to ask dumb questions to 😁 

What about

```peg
Document = names

WS = [ \r\n]+

fname = "bob" / "dan" / "charlie"

namelist = nl:(namelast ",")+ { return nl[0][0]; }
namelast = WS? fn:fname       { return fn; }

names = nl:namelist? na:namelast { return [].concat(nl, na); } 
```

**Comment by StoneCypher** [comment]
Or Ukrainian or whatever.  I apologize: I saw a Cyrillic name.  I shouldn't guess like that.  Getting that wrong can be offensive.

**Comment by Mingun** [comment]
Works, of course, but here we approach the question put in the headline issue -- a _simpler way_. Separated data are used quite often to have a separate syntax for them. In my practice this is the second place for which I use ranges, the first is repetitions with variable data (`peg = len:number someData|len|;` in my range syntax)

And yes, I'm from Russia. Don't worry, no offense

**Comment by StoneCypher** [comment]
> > Okay, this is starting to look like something that genuinely should have an operator, as an ease of use thing
> 
>  here we approach the question put in the headline issue -- a *simpler way*.

Yeah, I agree now.  It's just that my first wrong attempt looked really simple, and if it wasn't wrong, that would have been simple enough to not bother

but now that I see what's required, I agree
